File: C++/Problems/Problems/problem_3917.cpp
#include "problem_3917.h"

bool problem_3917::test() {
    std::vector nums { 1, 2, 3, 4 };

    std::vector expected { 2, 1, 1, 0 };

    auto result = countOppositeParity(nums);

    print1DVector(result);

    return expected == result;
}

std::vector<int> problem_3917::countOppositeParity(std::vector<int> &nums) {
    int n = nums.size();
    int even = 0;
    int odd = 0;
    std::vector<int> result(n);
    for (int i = n - 1; i >= 0; --i)
    {
        if (nums[i] % 2 == 0)
        {
            result[i] = odd;
            even++;
        }
        else
        {
            result[i] = even;
            odd++;
        }
    }

    return result;
}

File: Python/Problems/Problems/problem3917.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem3917(Problem):
    def test(self) -> bool:
        nums = [1, 2, 3, 4]

        expected = [2, 1, 1, 0]

        result = self.countOppositeParity(nums)

        print1DArray(result)

        return expected == result

    def countOppositeParity(self, nums: list[int]) -> list[int]:
        n = len(nums)
        even = 0
        odd = 0
        result = [0] * n
        for i in range(n - 1, -1, -1):
            if nums[i] % 2 == 0:
                result[i] = odd
                even += 1
            else:
                result[i] = even
                odd += 1

        return result